pkg/machine/machineset.go

```diff
@@ -0,0 +1,378 @@
+package machine
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/openshift-kni/eco-goinfra/pkg/clients"
+	"github.com/openshift-kni/eco-goinfra/pkg/msg"
+	machinev1beta1 "github.com/openshift/api/machine/v1beta1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+// MachineSetBuilder provides a struct for MachineSet object from the cluster and a MachineSet definition.
+type MachineSetBuilder struct {
+	// MachineSetBuilder definition. Used to create
+	// MachineSet object with minimum set of required elements.
+	Definition *machinev1beta1.MachineSet
+	// Created MachineSetBuilder object on the cluster.
+	Object *machinev1beta1.MachineSet
+	// api client to interact with the cluster.
+	apiClient *clients.Settings
+	// errorMsg is processed before MachineSetBuilder object is created.
+	errorMsg string
+}
+
+// NewMachineSetBuilderFromCopy returns an MachineSetBuilder struct from a copied MachineSet.
+func NewMachineSetBuilderFromCopy(
```

Comment:
@wabouhamad would it be easier to seperater the NewBuilder functions by provider type? For example, you can call this NewAWSMachineSetBuilder and then return a MachineSetBuilder with .Spec.Template.Spec.ProviderSpec.Value set to an AWSMachineProviderConfig I believe

Reply:
I have redesigned the code to auto-detect the public cloud based on discussions we had on slack.